RHOAIENG-51591: Removing support for LORA-distributed and 8bit quantization issue with unsloth. (#41)
red-hat-data-services/red-hat-ai-examples: examples/fine-tuning/lora/README.md

This notebook demonstrates how to use Training Hub's LoRA (Low-Rank Adaptation) and QLoRA capabilities for parameter-efficient fine-tuning. We'll train a model to convert natural language questions into SQL queries using the popular [sql-create-context](https://huggingface.co/datasets/b-mc2/sql-create-context) dataset.

> [!KNOWN-ISSUES]

> > - Lora relies on unsloth and it doesn't currently support multi node, so we are currently supporting interactive execution mode.

Reviewer: Maybe something like this? 
```suggestion
> - Multi-node distributed is not currently supported as LoRA via Training Hub relies on Unsloth which does not currently support it.
```

> > - QLORA currently does not support 8 bit quantization mode, only 4bit quantization is supported because of limitations of unsloth.

Reviewer: I think Unsloth does now support 8 bit quantization so maybe this could be reworded?
Something about, 4-bit being the most commonly used for QLoRA due to its superior memory efficiency and speed? What do you think?

> This notebook demonstrates how to use Training Hub's LoRA (Low-Rank Adaptation) and QLoRA capabilities for parameter-efficient fine-tuning. We'll train a model to convert natural language questions into SQL queries using the popular [sql-create-context](https://huggingface.co/datasets/b-mc2/sql-create-context) dataset.

Reviewer: This is kind of repeating the line above it, do we need both or could we amalgamate them?

## Note

> LORA/QLORA supports only following mode of execution:

Reviewer: ```suggestion
LORA/QLORA supports:
```

Reviewer: Please update this, using `only` is quite negative.

|------------|----------|-----|-----|--------|-------|
| CUDA PyTorch Python 3.12 | NVIDIA GPU training | 1× GPU | 4 cores | 32Gi | Recommended for faster training |

**Note**

- Interactive mode is recommended for smaller training jobs.

**Note**

> - Workbench GPU is optional but recommended for faster model evaluation

Reviewer 2: is it optional?
Can lora on CPU only? Also, will GPU just be for faster model eval or also for fine tune?

Author: Yes it will be faster for evaluation with GPUs. but if its a very small task a small latency for evaluation is not a big deal. its a tradeoff between the cost and amount of data they are evaluating with.

| CUDA PyTorch Python 3.12 | NVIDIA GPU training | 1× GPU | 4 cores | 32Gi | Recommended for faster training |

> [!NOTE]
> - Interactive mode is recommended for smaller training jobs.

> [!NOTE]

> > - This example was tested on 2 nodes x 2 GPUs provided by L40S however, it will work on smaller/larger configurations.

Reviewer: This is related to multi-node so should be updated

> - CPU and Memory requirements scale with batch size and model size. Above suit the example as it is.
> - Worker pods are configurable from the `client.create_job` call within the notebook.

> [!NOTE]
> - Workbench GPU is optional but recommended for faster model evaluation
> - Evaluation was performed on L40S GPU however, it will work on smaller/larger configurations.
> - Workbench resources and accelerator are configurable in `Create Workbench` view on RHOAI Platform

> [!NOTE]

> > - Storage can be created in `Create Workbench` view on RHOAI Platform, however, dynamic RWX provisioner is required to be configured prior to creating shared file storage in RHOAI.

Reviewer: Is this valid for interactive?

    > [!NOTE]

>     > Adding an accelerator (GPU) for the distributed use case is only needed to test the fine-tuned model from within the workbench so you can spare an accelerator if you plan to skip that step. An accelerator (GPU) is required in interactive mode as the training happens on the workbench pod.

Reviewer 2: same, if we are only interactive / local / th is GPU really an option or a must?

Reviewer 2: Ah I see it mentions further that for fine tuning interactive GPU is required. I would reword / remove the distributed parts from the doc!

  - Review the storage configuration and click "Create workbench":
    ![](./docs/04e.png)

- From "Workbenches" page, click on **_Open_** when the workbench you've just created becomes ready:
  ![](./docs/05.png)

**Important**

- By default:

>   - The example goes through distributed training on two nodes (2xL40/L40S) with two GPUs each (2x48GB). However, it can be tweaked to run on smaller configurations.

Reviewer 2: We removed distributed right?
